File: 2015/day_1/main.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

char *read_to_string(char *filename);
void panic(char *msg);

int main(int argc, char *argv[]) {
    printf("hello\n");

    char *filename = "input.txt";
    FILE *file = fopen(filename, "r");
    if (file == NULL) {
        printf("Error: file %s not found", filename);
        exit(-1);
    }

    int i = 0;
    int floor = 0;
    char ch;
    while (fscanf(file, "%c", &ch) != EOF) {
        if (ch == '\n') {
            continue;
        }

        if (ch == '(') {
            floor++;
        } else if (ch == ')') {
            floor--;
        }

        if (floor == -1) {
            printf("Santa enters the basement at position %d\n", i + 1);
            break;
        }
        i++;
    }
    fclose(file);
    return 0;
}

void panic(char *msg) {
    printf("%s\n", msg);
    exit(-1);
}

File: 2015/day_1/src/main.rs
use std::fs;

fn main() {
    let filename = "input.txt";
    let file_contents = match fs::read_to_string(filename) {
        Ok(v) => v,
        Err(_) => panic!("could not read file"),
    };

    if let Err(e) = run(&file_contents) {
        panic!("Error {}", e);
    }
}

fn run(file_contents: &str) -> Result<(), &'static str> {
    let mut floor = 0;
    for (i, c) in file_contents.chars().enumerate() {
        floor += if c == '(' { 1 } else { -1 };
        if floor == -1 {
            println!("Santa enters the basement at position {}", i + 1);
            return Ok(());
        }
    }

    println!("Floor: {floor}");

    Ok(())
}
